File: receipt-analyzer-service/src/main/java/pl/madzierski/daniel/app/file_group/FileGroupProvider.java
package pl.madzierski.daniel.app.file_group;

import lombok.AllArgsConstructor;
import org.springframework.http.MediaType;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;
import pl.madzierski.daniel.app.file_group.file.FileEntity;
import pl.madzierski.daniel.app.file_group.file.FileProvider;
import pl.madzierski.daniel.app.receipt.ReceiptEntity;
import pl.madzierski.daniel.exception.AppRuntimeException;
import pl.madzierski.daniel.exception.AppRuntimeExceptionMessages;
import pl.madzierski.daniel.security.SecurityUtils;

import java.util.HashSet;

@Service
@AllArgsConstructor
public class FileGroupProvider {

    private final FileGroupRepository fileGroupRepository;
    private final FileProvider fileProvider;

    @Transactional
    public FileGroupEntity saveReceiptFile(ReceiptEntity receipt, MultipartFile file) {
        FileType fileType = FileType.invoke(file.getContentType());
        FileGroupEntity fileGroupEntity = new FileGroupEntity(fileType, receipt, true);
        fileGroupEntity = fileGroupRepository.save(fileGroupEntity);

        FileEntity fileEntity = new FileEntity(fileGroupEntity, null, null, 0);
        fileGroupEntity.addFile(fileEntity);
        fileEntity = fileProvider.save(fileEntity);

        String pathInString = createPath(SecurityUtils.getCurrentUserSub(), fileEntity, fileGroupEntity, fileType.getExtension(), receipt);

        fileProvider.saveFile(pathInString, file);
        fileEntity.setPath(pathInString);
        fileProvider.save(fileEntity);

        return fileGroupRepository.save(fileGroupEntity);
    }


    public String createPath(
            String currentUserSub,
            FileEntity file,
            FileGroupEntity fileEntity,
            String fileExtension,
            ReceiptEntity receipt
    ) {
        return String.format("/app/uploads/user/%s/receipts/%s/file_group/%s/file/%s.%s",
                currentUserSub, receipt.getId(), fileEntity.getId(), file.getId(), fileExtension);
    }

}

File: receipt-analyzer-service/src/main/kotlin/pl/madzierski/daniel/app/file_group/FileGroupProvider.kt
package pl.madzierski.daniel.app.file_group

import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.multipart.MultipartFile
import pl.madzierski.daniel.app.file_group.file.FileEntity
import pl.madzierski.daniel.app.file_group.file.FileProvider
import pl.madzierski.daniel.app.receipt.ReceiptEntity
import pl.madzierski.daniel.exception.AppRuntimeException
import pl.madzierski.daniel.exception.AppRuntimeExceptionMessages
import pl.madzierski.daniel.security.SecurityUtils

@Service
class FileGroupProvider(private val fileGroupRepository: FileGroupRepository, private val fileProvider: FileProvider) {

    @Transactional
    fun saveReceiptFile(receipt: ReceiptEntity, file: MultipartFile): FileGroupEntity {
        var fileGroupEntity = FileGroupEntity(FileType.PDF, receipt, true, mutableSetOf())
        fileGroupEntity = fileGroupRepository.save(fileGroupEntity)
        var fileEntity = FileEntity(fileGroupEntity, null, null, 0)
        fileGroupEntity.files.add(fileEntity)
        fileEntity = fileProvider.save(fileEntity)
        val fileExtension = getExtension(file.contentType)
        val pathInString =
            createPath(SecurityUtils.getCurrentUserSub(), fileEntity, fileGroupEntity, fileExtension, receipt)
        fileProvider.saveFile(pathInString, file)
        fileEntity.path = pathInString
        fileProvider.save(fileEntity)
        val save = fileGroupRepository.save(fileGroupEntity)
        return save
    }


    private fun getExtension(contentType: String?): String {
        return when (contentType) {
            MediaType.APPLICATION_PDF_VALUE -> "pdf"
            else -> throw AppRuntimeException(AppRuntimeExceptionMessages.UNHANDLED_MEDIA_TYPE)
        }
    }

    fun createPath(
        currentUserSub: String,
        file: FileEntity,
        fileEntity: FileGroupEntity,
        fileExtension: String,
        receipt: ReceiptEntity
    ): String =
        "/app/uploads/user/$currentUserSub/receipts/${receipt.id}/file_group/${fileEntity.id}/file/${file.id}.${fileExtension}"

    fun save(fileGroupEntity: FileGroupEntity): FileGroupEntity = fileGroupRepository.save(fileGroupEntity)

    fun getOriginalPdf(receiptId: String): String {
        return fileGroupRepository.findFirstOriginalPdf(receiptId)
    }

}